src/nnet3/nnet-descriptor.h

```diff
@@ -188,10 +203,21 @@ class SwitchingForwardingDescriptor: public ForwardingDescriptor {
   // that this descriptor may access.
   virtual void GetNodeDependencies(std::vector<int32> *node_indexes) const;
 
+  virtual std::string Type() const { return "SwitchingForwardingDescriptor"; }
+
   // takes ownership of items in src.
   SwitchingForwardingDescriptor(std::vector<ForwardingDescriptor*> &src):
       src_(src) { }
   virtual ~SwitchingForwardingDescriptor() { DeletePointers(&src_); }
+
+  // defined in nnet-utils.h
+  friend void TraceIntoForwardingDescriptorForOffsets(const Nnet &nnet,
+                                          const ForwardingDescriptor
+                                          &this_descriptor,
+                                          const std::vector<std::string>
+                                          &recurrent_node_names,
```

Comment:
recurrent_node_names and recurrent_offsets can just be
node_names and offsets, as these methods do not assume anything about recurrence (right ?)